core-bundle/src/Controller/BackendPreviewSwitchController.php

```diff
@@ -0,0 +1,149 @@
+        $this->connection                   = $connection;
+    }
+
+    /**
+     * @Route("/_contao/preview_switch", name="contao_backend_preview_switch")
+     */
+    public function __invoke(Request $request): Response
+    {
+        $this->contaoFramework->initialize(false);
+
+        $user = BackendUser::getInstance();
+        if (null === $user || !$request->isXmlHttpRequest()) {
+            throw new PageNotFoundException();
+        }
+
+        if ('tl_switch' === $request->request->get('FORM_SUBMIT')) {
+            $this->authenticatePreview($user, $request);
+
+            return Response::create();
+        }
+
+        if ('datalist_members' === $request->request->get('FORM_SUBMIT')) {
+            $data = $this->getMembersDataList($user, $request);
+
+            return JsonResponse::create($data);
+        }
+
+        return Response::create('', 404);
+    }
+
+    private function authenticatePreview(BackendUser $user, Request $request): void
+    {
+        $canSwitchUser    = $this->isAllowedToAccessMembers($user);
+        $frontendUsername = $this->tokenChecker->getFrontendUsername();
+        $showUnpublished  = 'hide' !== $request->request->get('unpublished');
+
+        if ($canSwitchUser) {
+            $frontendUsername = $request->request->get('user') ?: null;
+        }
+
+        if (null !== $frontendUsername) {
+            $this->frontendPreviewAuthenticator->authenticateFrontendUser($frontendUsername, $showUnpublished);
+        } else {
+            $this->frontendPreviewAuthenticator->authenticateFrontendGuest($showUnpublished);
+        }
+    }
+
+    private function getMembersDataList(BackendUser $user, Request $request): array
+    {
+        $andWhereGroups = '';
+
+        if (!$this->isAllowedToAccessMembers($user)) {
+            return [];
+        }
+
+        if (!$user->isAdmin) {
+            $groups = array_map(
+                static function ($groupId) {
+                    return '%"' . (int) $groupId . '"%';
+                },
+                $user->amg
+            );
+
+            $andWhereGroups = "AND (groups LIKE '" . implode("' OR GROUPS LIKE '", $groups) . "')";
+        }
+
+        $time = Date::floorToMinute();
+
+        // Get the active front end users
+        $result = $this->connection->executeQuery(
+            sprintf(
+                <<<SQL
+SELECT username 
+FROM tl_member 
+WHERE username LIKE ?
+%s 
+AND login='1' AND disable!='1' AND (start='' OR start<='%s') AND (stop='' OR stop>'%d')
+ORDER BY username
+SQL
+                ,
+                $andWhereGroups,
+                $time,
+                $time + 60
+            ),
+            [str_replace('%', '', $request->request->get('value')) . '%']
+        );
+
+        return $result->fetchAll(FetchMode::COLUMN);
+    }
+
+    private function isAllowedToAccessMembers(BackendUser $user): bool
+    {
+        return ($user->isAdmin || (!empty($user->amg) && \is_array($user->amg)));
```

Comment:
Should use security voters (at some point). Maybe we can even come up with a role for that?

Reply:
There is `$this->security->isGranted('contao_user.amg' /* , $subject = null */,);` available but atm you MUST pass an array of member groups as second parameter. Therefore, we cannot use the existing `BackendAccessVoter` to check if a user has any member groups allowed.